crates/services/p2p/src/service.rs

```diff
@@ -107,65 +118,271 @@ impl Debug for TaskRequest {
+    }
+
+    fn next_event(&mut self) -> BoxFuture<'_, Option<FuelP2PEvent>> {
+        Box::pin(self.next_event())
+    }
+
+    fn publish_message(
+        &mut self,
+        message: GossipsubBroadcastRequest,
+    ) -> anyhow::Result<()> {
+        self.publish_message(message)?;
+        Ok(())
+    }
+
+    fn send_request_msg(
+        &mut self,
+        peer_id: Option<PeerId>,
+        request_msg: RequestMessage,
+        channel_item: ResponseChannelItem,
+    ) -> anyhow::Result<()> {
+        self.send_request_msg(peer_id, request_msg, channel_item)?;
+        Ok(())
+    }
+
+    fn send_response_msg(
+        &mut self,
+        request_id: RequestId,
+        message: OutboundResponse,
+    ) -> anyhow::Result<()> {
+        self.send_response_msg(request_id, message)?;
+        Ok(())
+    }
+
+    fn report_message(
+        &mut self,
+        message: GossipsubMessageInfo,
+        acceptance: GossipsubMessageAcceptance,
+    ) -> anyhow::Result<()> {
+        report_message(self, message, acceptance);
+        Ok(())
+    }
+
+    fn report_peer(
+        &mut self,
+        peer_id: PeerId,
+        score: AppScore,
+        reporting_service: &str,
+    ) -> anyhow::Result<()> {
+        self.report_peer(peer_id, score, reporting_service);
+        Ok(())
+    }
+
+    fn update_block_height(&mut self, height: BlockHeight) -> anyhow::Result<()> {
+        self.update_block_height(height);
+        Ok(())
+    }
+}
+
+pub trait Broadcast: Send {
+    fn report_peer(
+        &self,
+        peer_id: FuelPeerId,
+        report: HeartBeatPeerReportReason,
+        reporting_service: &'static str,
+    ) -> anyhow::Result<()>;
+
+    fn block_height_broadcast(
+        &self,
+        block_height_data: BlockHeightHeartbeatData,
+    ) -> anyhow::Result<()>;
+
+    fn tx_broadcast(&self, transaction: TransactionGossipData) -> anyhow::Result<()>;
+}
+
+impl Broadcast for SharedState {
+    fn report_peer(
+        &self,
+        peer_id: FuelPeerId,
+        report: HeartBeatPeerReportReason,
+        reporting_service: &'static str,
+    ) -> anyhow::Result<()> {
+        self.report_peer(peer_id, report, reporting_service)
+    }
+
+    fn block_height_broadcast(
+        &self,
+        block_height_data: BlockHeightHeartbeatData,
+    ) -> anyhow::Result<()> {
+        self.block_height_broadcast.send(block_height_data)?;
+        Ok(())
+    }
+
+    fn tx_broadcast(&self, transaction: TransactionGossipData) -> anyhow::Result<()> {
+        self.tx_broadcast.send(transaction)?;
+        Ok(())
+    }
+}
+
 /// Orchestrates various p2p-related events between the inner `P2pService`
 /// and the top level `NetworkService`.
-pub struct Task<D> {
-    p2p_service: FuelP2PService<PostcardCodec>,
+pub struct Task<P, D, B> {
+    p2p_service: P,
     db: Arc<D>,
     next_block_height: BoxStream<BlockHeight>,
     /// Receive internal Task Requests
     request_receiver: mpsc::Receiver<TaskRequest>,
-    shared: SharedState,
```

Comment:
minor nit: `shared` is conventional for most of our tasks in fuel core

Reply:
It is still represented in the Task trait assoc type. I'm not a fan of having a field called `shared` because it doesn't communicate at all what it's for and why it's grouped. `shared` implies to me that it's for misc data and anyone extending it might just throw it in there. Whereas the trait assoc type communicates that it would probably be shared with other Tasks. 

I was hoping to move away from that name :)

Reply:
In this case I think it's fine, since this task is just mimicking the form, but not rigidly adhering to the standardized task framework from `fuel-core-services`. However, changing it in the broader scope will require a more comprehensive plan to improve.